feature_extractor.py

```diff
@@ -0,0 +1,292 @@
+        """
+        assert inputs.shape[1] == self.height
+        assert inputs.shape[2] == self.width
+
+        inputs = inputs[:, None, :, :].float()
+
+        x = coordinates[:, 0]
+        y = coordinates[:, 1]
+
+        x_shift = 2 / (self.width - 1) * (x.float() - self.width // 2)
+        y_shift = 2 / (self.height - 1) * (y.float() - self.height // 2)
+
+        grid = torch.stack(
+            [
+                self.width_grid[None, :, :] + x_shift[:, None, None],
+                self.height_grid[None, :, :] + y_shift[:, None, None],
+            ],
+            dim=3,
+        )
+
+        # TODO: only cast to int if original tensor was int
+        return (
+            torch.round(F.grid_sample(inputs, grid, align_corners=True))
+            .squeeze(1)
+            .long()
+        )
+
+
+class NetHackNet(nn.Module):
+    def __init__(
+        self,
+        observation_shape,
+        num_actions,
+        use_lstm,
+        embedding_dim=32,
+        crop_dim=9,
+        num_layers=5,
+    ):
+        super(NetHackNet, self).__init__()
+
+        self.glyph_shape = observation_shape["glyphs"].shape
+        self.blstats_size = observation_shape["blstats"].shape[0]
+
+        self.num_actions = num_actions
+        self.use_lstm = use_lstm
+
+        self.H = self.glyph_shape[0]
+        self.W = self.glyph_shape[1]
+
+        self.k_dim = embedding_dim
+        self.h_dim = 512
+
+        self.crop_dim = crop_dim
+
+        self.crop = Crop(self.H, self.W, self.crop_dim, self.crop_dim)
+
+        self.embed = nn.Embedding(nethack.MAX_GLYPH, self.k_dim)
+
+        K = embedding_dim  # number of input filters
+        F = 3  # filter dimensions
+        S = 1  # stride
+        P = 1  # padding
+        M = 16  # number of intermediate filters
+        Y = 8  # number of output filters
+        L = num_layers  # number of convnet layers
+
+        in_channels = [K] + [M] * (L - 1)
+        out_channels = [M] * (L - 1) + [Y]
+
+        def interleave(xs, ys):
+            return [val for pair in zip(xs, ys) for val in pair]
+
+        conv_extract = [
+            nn.Conv2d(
+                in_channels=in_channels[i],
+                out_channels=out_channels[i],
+                kernel_size=(F, F),
+                stride=S,
+                padding=P,
+            )
+            for i in range(L)
+        ]
+
+        self.extract_representation = nn.Sequential(
+            *interleave(conv_extract, [nn.ELU()] * len(conv_extract))
+        )
+
+        # CNN crop model.
+        conv_extract_crop = [
+            nn.Conv2d(
+                in_channels=in_channels[i],
+                out_channels=out_channels[i],
+                kernel_size=(F, F),
+                stride=S,
+                padding=P,
+            )
+            for i in range(L)
+        ]
+
+        self.extract_crop_representation = nn.Sequential(
+            *interleave(conv_extract_crop, [nn.ELU()] * len(conv_extract))
```

Comment:
My general preference is to make a _block_, instead of doing something like this, which feels like an awkward workaround.
`class Conv2dWithActivation([...], activation_fn='...')`
But this is just my preference.

Reply:
-> FB